File: src/main/java/org/taktik/icure/logic/TarificationLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.Tarification;

public interface TarificationLogic {
	Tarification get(String id);

	Tarification get(@NotNull String type, @NotNull String tarification, @NotNull String version);

	List<Tarification> get(List<String> ids);

	Tarification create(Tarification tarification);

	Tarification modify(Tarification tarification) throws Exception;

	List<Tarification> findTarificationsBy(String type, String tarification, String version);

	List<Tarification> findTarificationsBy(String region, String type, String tarification, String version);

	PaginatedList<Tarification> findTarificationsBy(String region, String type, String tarification, String version, PaginationOffset paginationOffset);

	PaginatedList<Tarification> findTarificationsByLabel(String region, String language, String label, PaginationOffset paginationOffset);

	PaginatedList<Tarification> findTarificationsByLabel(String region, String language, String type, String label, PaginationOffset paginationOffset);

	Tarification getOrCreateTarification(String type, String tarification);
}

File: src/main/kotlin/org/taktik/icure/asynclogic/TarificationLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.Tarification

interface TarificationLogic {
    suspend fun getTarification(id: String): Tarification?
    suspend fun getTarification(type: String, tarification: String, version: String): Tarification?
    fun getTarifications(ids: List<String>): Flow<Tarification>
    suspend fun createTarification(tarification: Tarification): Tarification?
    suspend fun modifyTarification(tarification: Tarification): Tarification?

    fun findTarificationsBy(type: String?, tarification: String?, version: String?): Flow<Tarification>
    fun findTarificationsBy(region: String?, type: String?, tarification: String?, version: String?): Flow<Tarification>
    fun findTarificationsBy(
        region: String?,
        type: String?,
        tarification: String?,
        version: String?,
        paginationOffset: PaginationOffset<List<String?>>
    ): Flow<ViewQueryResultEvent>

    fun findTarificationsByLabel(
        region: String?,
        language: String?,
        label: String?,
        paginationOffset: PaginationOffset<List<String?>>
    ): Flow<ViewQueryResultEvent>
    fun findTarificationsByLabel(region: String?, language: String?, type: String?, label: String?, paginationOffset: PaginationOffset<List<String?>>): Flow<ViewQueryResultEvent>
    suspend fun getOrCreateTarification(type: String, tarification: String): Tarification?
}
